blog/_posts/2021-07-08-portforwarding-is-awesome.md

```diff
@@ -0,0 +1,292 @@
+```
+
+Tilt will periodically check that the portforward is running. Other tools for
+managing multiple services (e.g., tmux) will work well too!
+
+## Outside the Box: A port-forward is A Special Dynamic Server!
+
+But most Tilt users don't think of the port-forward as a separate server. We
+think about it as a property of the primary server! In Tilt, you specify it as a
+property of a Kubernetes resource:
+
+```python
+k8s_resource('example-go', port_forwards=[port_forward(8080, 8000)])
+```
+
+If we delete the primary server, we want the port-forward to go away.
+
+If we delete a pod and replace it with a new one, we want the port-forward to send
+traffic to the new pod.
+
+How does Tilt accomplish this?
+
+Tilt has two important tools to stitch this together: 
+the [PortForward](https://api.tilt.dev/kubernetes/port-forward-v1alpha1.html) object 
+and the [KubernetesDiscovery](https://api.tilt.dev/kubernetes/kubernetes-discovery-v1alpha1.html) object.
+
+PortForward tracks all the port-forwards that Tilt is managing. If I'm running
+[the example-go project](https://github.com/tilt-dev/tilt-example-go), I can
+inspect them like this:
+
+```
+$ tilt get portforwards
+NAME                                     CREATED AT
+example-go-example-go-5cc87c754d-9ngn4   2021-07-07T18:36:05Z
+
+$ tilt describe portforwards
+Name:         example-go-example-go-5cc87c754d-9ngn4
+...
+API Version:  tilt.dev/v1alpha1
+Kind:         PortForward
+Metadata:
+  Creation Timestamp:  2021-07-07T18:36:05Z
+  ...
+Spec:
+  Forwards:
+    Container Port:  8000
+    Host:            localhost
+    Local Port:      8000
+  Namespace:         default
+  Pod Name:          example-go-5cc87c754d-9ngn4
+Status:
+  Forward Statuses:
+    Addresses:
+      127.0.0.1
+      ::1
+    Container Port:  8000
+    Local Port:      8000
+    Started At:      2021-07-07T18:36:06.022789Z
+```
+
+This lets you see what pods Tilt is connecting to and when it started port-forwarding.
+
+But each PortForward object only works for a single pod.
+
+Tilt has a second API object, KubernetesDiscovery, which tracks which pods belong to which
+resource.
+
+```
+$ tilt get kubernetesdiscovery
+NAME         CREATED AT
+example-go   2021-07-07T18:35:51Z
+$ tilt describe kubernetesdiscovery
+Name:         example-go
+...
+Spec:
+  ...
+  Port Forward Template Spec:
+    Forwards:
+      Container Port:  0
+      Host:            localhost
+      Local Port:      8000
+  Watches:
+    Name:       example-go
+    Namespace:  default
+    UID:        bcb9e5bd-bd8c-4683-87b1-b7b7b7276c7f
+Status:
+  Monitor Start Time:  2021-07-07T18:36:04.509869Z
+  Pods:
+    ...
+    Name:                    example-go-5cc87c754d-9ngn4
+    ...
+```
+
+The KubernetesDiscovery object contains a port-forward template. As pods are
+created, updated, and deleted, this object will select the "best" pod for
+portforwarding, and sets up a PortForward object to manage the connection.
+
+## Why am I telling you this??? 
```

Comment:
I think this section is good from an overview/philosophical perspective but going back to my first comment up top, what do readers gain from this exactly in practical terms?

"It's 2AM, I'm tired, and not very smart, connect the dots for me, what should I use this knowledge for?"

Reply:
ya, agreed, what do you think of the new intro?